ServerCore/SyncController.cs

```diff
@@ -0,0 +1,534 @@
+                // for them and don't count them.
+
+                if (solvedPuzzle.Event.ID != eventId) {
+                    continue;
+                }
+
+                // If the request is asking whether certain puzzles are solved, check if it's
+                // in the set and, if so, put it in the list of solved puzzles to inform the
+                // requester of.
+                    
+                if (queryPuzzleIdSet != null && queryPuzzleIdSet.Contains(solvedPuzzle.ID)) {
+                    solvedPuzzles.Add(solvedPuzzle.ID);
+                }
+
+                // When counting solves, only count puzzles if they're not in the group that doesn't
+                // count toward the solve count, and only if they're worth at least 10 points.
+
+                if (groupExcludedFromSolveCount == null || solvedPuzzle.Group != groupExcludedFromSolveCount) {
+                    if (solvedPuzzle.SolveValue >= 10) {
+                        maxSolveCount += 1;
+                    }
+                }
+
+                // If the user solved a cheat code, treat it as a solve count of 1000.
+
+                if (solvedPuzzle.IsCheatCode) {
+                    maxSolveCount += 1000;
+                }
+            }
+
+            // If the requester is asking for puzzle pieces (by setting request.MinSolveCount != null)
+            // and if there are pieces of the puzzle that the requester has now earned but hasn't
+            // yet received, because the maximum solve count they've earned is at least as high
+            // as the minimum solve count the requester is asking for, then return those pieces.
+            // Note that request.MinSolveCount is the minimum solve count of tokens the requester
+            // *hasn't* seen yet.
+
+            if (request.MinSolveCount != null && maxSolveCount >= request.MinSolveCount) {
+                List<Piece> pieces = await (from piece in context.Pieces
+                                            where piece.ProgressLevel >= request.MinSolveCount && piece.ProgressLevel <= maxSolveCount
+                                            select piece).ToListAsync();
+                response.SetMinAndMaxSolveCountAndPieces(request.MinSolveCount.Value, maxSolveCount, pieces);
+            }
+
+            // If we found some solved puzzles in request.QueryPuzzleIds, return those in the
+            // response.
+
+            if (solvedPuzzles.Count > 0) {
+                response.SetSolvedPuzzles(solvedPuzzles);
+            }
+        }
+
+        /// <summary>
+        ///   This routine stores in the database any annotations the requester has uploaded.
+        /// </summary>
+        private async Task StoreAnnotations(DecodedSyncRequest request, SyncResponse response, int puzzleId, int teamId)
+        {
+            if (request.AnnotationRequests == null) {
+                return;
+            }
+
+            foreach (var annotationRequest in request.AnnotationRequests) {
```

Comment:
If you wrap the whole thing in a transaction does that help with the concurrency problems and the special SQL etc.?

Reply:
Unfortunately, no, because SQL's default setting is read committed, not strict serialization.  If you have the database set up for strict serialization, though, let me know and I'd be happy to use a transaction instead!

Reply:
As if I had any clue :-)